atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2509,38 +2507,99 @@ private void logCommitLockTenureExceeded(
+                        collectCellLocks(lockDescriptorBuilder, lockDescriptorToChangeMetadataBuilder, tableRef);
             }
-
             if (conflictHandler.lockRowsForConflicts()) {
-                Cell lastCell = null;
-                for (Cell cell : getLocalWrites(tableRef).keySet()) {
-                    if (lastCell == null || !Arrays.equals(lastCell.getRowName(), cell.getRowName())) {
-                        result.add(AtlasRowLockDescriptor.of(tableRef.getQualifiedName(), cell.getRowName()));
-                        rowLockCount++;
-                    }
-                    lastCell = cell;
-                }
+                rowLockCount += collectRowLocks(lockDescriptorBuilder, lockDescriptorToChangeMetadataBuilder, tableRef);
             }
         }
-        result.add(AtlasRowLockDescriptor.of(
+        lockDescriptorBuilder.add(AtlasRowLockDescriptor.of(
                 TransactionConstants.TRANSACTION_TABLE.getQualifiedName(),
                 TransactionConstants.getValueForTimestamp(getStartTimestamp())));
         cellCommitLocksRequested = cellLockCount;
         rowCommitLocksRequested = rowLockCount + 1;
-        return result;
+        Map<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadata =
+                lockDescriptorToChangeMetadataBuilder.buildOrThrow();
+        return LocksAndMetadata.of(
+                lockDescriptorBuilder.build(),
+                // For now, lock request metadata only consists of change metadata. If it is absent, we can save
+                // computation by not doing an index encoding
+                lockDescriptorToChangeMetadata.isEmpty()
+                        ? Optional.empty()
+                        : Optional.of(LockRequestMetadata.of(lockDescriptorToChangeMetadata)));
+    }
+
+    /**
+     * Returns the number of lock descriptors that were added.
+     */
+    private Long collectCellLocks(
+            ImmutableSet.Builder<LockDescriptor> lockDescriptorBuilder,
+            ImmutableMap.Builder<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadataBuilder,
+            TableReference tableRef) {
+        long lockCount = 0;
+        Map<Cell, ChangeMetadata> changeMetadataForWrites =
+                localWriteBuffer.getChangeMetadataForWritesToTable(tableRef);
+        for (Cell cell : localWriteBuffer.getLocalWritesForTable(tableRef).keySet()) {
+            LockDescriptor lockDescriptor =
+                    AtlasCellLockDescriptor.of(tableRef.getQualifiedName(), cell.getRowName(), cell.getColumnName());
+            lockDescriptorBuilder.add(lockDescriptor);
+            lockCount++;
+            if (changeMetadataForWrites.containsKey(cell)) {
+                lockDescriptorToChangeMetadataBuilder.put(lockDescriptor, changeMetadataForWrites.get(cell));
+            }
+        }
+        return lockCount;
+    }
+
+    /**
+     * Returns the number of lock descriptors that were added.
+     */
+    private Long collectRowLocks(
```

Comment:
Should we just move this into `LocalWriteBuffer` ?

Reply:
(Ideally we would move this to it's own utility, but a bit tricky as we need the conflict handler...)